source/adapters/level_zero/v2/event_provider_normal.hpp

```diff
@@ -0,0 +1,67 @@
+//===--------- event_provider_normal.hpp - Level Zero Adapter -------------===//
+//
+// Copyright (C) 2024 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include <deque>
+#include <memory>
+#include <mutex>
+#include <stack>
+
+#include <unordered_map>
+#include <ur/ur.hpp>
+#include <ur_api.h>
+#include <ze_api.h>
+
+#include "../device.hpp"
+#include "common.hpp"
+#include "event.hpp"
+
+namespace v2 {
+
+enum queue_type {
+  QUEUE_REGULAR,
+  QUEUE_IMMEDIATE,
+};
+
+class provider_pool {
+public:
+  provider_pool(ur_context_handle_t, ur_device_handle_t, event_type,
+                queue_type);
+  ~provider_pool();
+
+  event_borrowed allocate();
+  size_t nfree() const;
+
+private:
+  ze_event_pool_handle_t Pool;
```

Comment:
I would prefer to have some raii wrappers around this (any L0 handle tbh) so that you don't have to manually call destroy in the dtor. Something like this (similar to what we have in tests): https://github.com/igchor/unified-runtime/commit/e775e85da47d29a13be81d1dc63aa3a4019371b3

Reply:
Let's refactor this once your PR comes in and we can use the wrapper. I added a TODO.